mxcubecore/protocols_config.py

```diff
@@ -0,0 +1,135 @@
+        #
+        for channel_name, channel_config in device_config.get_channels():
+            attrs = dict(type="tango", name=channel_name, tangoname=device_name)
+            if channel_config.poll:
+                attrs["polling"] = channel_config.poll
+
+            hwobj.add_channel(attrs, channel_config.attribute)
+
+    tango_cfg = TangoConfig.parse_obj(tango_config)
+    for device_name, device_config in tango_cfg.get_tango_devices():
+        setup_tango_device(device_name, device_config)
+
+
+def _setup_exporter_commands_channels(hwobj: HardwareObject, exporter_config: dict):
+    from mxcubecore.model.protocols.exporter import ExporterConfig, Address
+
+    def setup_address(address: str, address_config: Address):
+        #
+        # set-up commands
+        #
+        for command_name, command_config in address_config.get_commands():
+            attrs = dict(
+                type="exporter", exporter_address=address, name=command_config.name
+            )
+            hwobj.add_command(attrs, command_name)
+
+        #
+        # set-up channels
+        #
+        for channel_name, channel_config in address_config.get_channels():
+            attrs = dict(type="exporter", exporter_address=address, name=channel_name)
+            hwobj.add_channel(attrs, channel_config.attribute)
+
+    exp_cfg = ExporterConfig.parse_obj(exporter_config)
+    for address, address_config in exp_cfg.get_addresses():
+        setup_address(address, address_config)
+
+
+def _setup_epics_channels(hwobj: HardwareObject, epics_config: dict):
+    from mxcubecore.model.protocols.epics import EpicsConfig, Prefix
+
+    def setup_prefix(prefix: str, prefix_config: Prefix):
+        #
+        # set-up channels
+        #
+        for channel_name, channel_config in prefix_config.get_channels():
+            attrs = dict(type="epics", name=channel_name)
+            if channel_config.poll:
+                attrs["polling"] = channel_config.poll
+
+            pv_name = f"{prefix}{channel_config.suffix}"
+            hwobj.add_channel(attrs, pv_name)
+
+    epics_cfg = EpicsConfig.parse_obj(epics_config)
+    for prefix, prefix_config in epics_cfg.get_prefixes():
+        setup_prefix(prefix, prefix_config)
+
+
+def _protocol_handles():
+    return {
+        "tango": _setup_tango_commands_channels,
+        "exporter": _setup_exporter_commands_channels,
+        "epics": _setup_epics_channels,
+    }
+
+
+def _get_protocol_names() -> Iterable[str]:
+    """
+    get names of all supported protocols
+    """
+    return _protocol_handles().keys()
+
+
+def _get_protocol_handler(protocol_name: str) -> Callable:
+    """
+    get the callable that will set up commands and channels for a specific protocol
+    """
+    return _protocol_handles()[protocol_name]
+
+
+def _setup_protocol(hwobj: HardwareObject, config: dict, protocol: str):
+    """
+    add the Command and Channel objects configured in the specified protocol section
+
+    parameters:
```

Comment:
I think the sections are case-sensitive.

```suggestion
    Parameters:
```

Reply:
works fine with lower case `p`:

![image](https://github.com/user-attachments/assets/04a6d6c7-5cc5-4c64-8091-67c6c7f808c1)